File: Leetcode/Java/0947. Most Stones Removed with Same Row or Column/src/Solution.java




import java.util.*;

public class Solution {
	public class MergeFind {
		int roots[];
		int ranks[];
		MergeFind(int numberOfNodes) {
			this.roots = IntStream.range(0, numberOfNodes + 10).toArray();
			this.ranks = new int[numberOfNodes + 10];
		}

		int find(int i) {
			if(i != roots[i]) { roots[i] = find(roots[i]); }
			return roots[i];
		}

		void merge(int u, int v) {
			final int uRoot = find(u), vRoot = find(v);
			if(uRoot != vRoot) {
				final int uRank = ranks[uRoot], vRank = ranks[vRoot];
				if(uRank < vRank) {
					roots[uRoot] = vRoot;					
				} else {
					roots[vRoot] = uRoot;
					ranks[uRoot] = ranks[uRoot] + 1;
				}
			}
		}

		int get(int[][] stones) {
			Set<Integer> seen = new HashSet<>();
			for(int[] stone: stones) {
				seen.add(find(stone[0]));
			}
			return stones.length - seen.size();
		}
	}

    public int removeStones(int[][] stones) {
		final int maxVertex = Arrays.stream(stones).flatMapToInt(Arrays::stream).max().getAsInt();		
		MergeFind ufo = new MergeFind(maxVertex + maxVertex + 1);
		for(int[] stone: stones) { ufo.merge(stone[0], stone[1] + maxVertex + 1); }
		
		return ufo.get(stones);
    }
}

File: Leetcode/Scala/0947. Most Stones Removed with Same Row or Column/Solution.scala




import scala.collection.mutable.{Set, ArrayBuffer}

class MergeFind private (val roots: Array[Int], val ranks: Array[Int]) {
    def this(numberOfElements: Int) = {
        this(Array.tabulate(numberOfElements + 10)(identity), new Array[Int](numberOfElements + 10))
    }

    def find(index: Int): Int = {
        if (roots(index) == index) index else {
            roots.update(index, find(roots(index)))
            roots(index)
        }
    }

    def merge(u: Int, v: Int): Unit = {
        val (uRoot, vRoot) = (find(u), find(v))
        if (uRoot != vRoot) {
            val (uRank, vRank) = (ranks(uRoot), ranks(vRoot))
            if (uRank < vRank) {
                roots.update(uRoot, vRoot)
            } else {
                roots.update(vRoot, uRoot)
                ranks.update(uRoot, ranks(vRoot) + 1)
            }
        }
    }

    def uniqueRoots(edges: Array[Array[Int]]): Int = {
        edges.length -
        edges.foldLeft(scala.collection.mutable.Set[Int]()) { case (seen, edge) =>
            seen.addOne(find(edge.head)); seen
        }.size
    }
}

object MergeFind {
    def apply(edges: Array[Array[Int]]): Int = {
        val mergeFind = new MergeFind(edges.flatten.max + edges.flatten.max + 1)
        edges.foreach { edge => mergeFind.merge(edge.head, edge.last + edges.flatten.max + 1) }
        mergeFind.uniqueRoots(edges)
    }
}

object Solution {
    def removeStones(stones: Array[Array[Int]]): Int = {
        MergeFind(stones)
    }
}
